crates/sui-core/src/validator_client_monitor/stats.rs

```diff
@@ -0,0 +1,246 @@
+
+/// Complete client-observed statistics for validator interactions.
+///
+/// This struct maintains client-side metrics for all validators in the network,
+/// including reliability scores, latency measurements, and failure tracking
+/// as observed from the client's perspective. It uses exponential moving averages (EMA)
+/// to smooth out transient spikes while still responding to sustained changes.
+#[derive(Debug, Clone)]
+pub struct ClientObservedStats {
+    /// Per-validator statistics mapping validator names to their client-observed metrics
+    pub validator_stats: HashMap<AuthorityName, ValidatorClientStats>,
+    /// Global statistics used for normalization and comparison
+    pub global_stats: GlobalStats,
+    /// Configuration parameters for scoring and exclusion policies
+    pub config: ValidatorClientMonitorConfig,
+}
+
+/// Client-observed stats for a single validator.
+///
+/// Tracks reliability, latency, and failure patterns for a specific validator
+/// as observed from the client's perspective. Uses exponential moving averages
+/// to smooth measurements while maintaining responsiveness to changes.
+#[derive(Debug, Clone)]
+pub struct ValidatorClientStats {
+    /// Exponential moving average of success rate (0.0 to 1.0)
+    pub reliability: DecayMovingAverage,
+    /// EMA latencies for each operation type (Submit, Effects, HealthCheck)
+    pub average_latencies: HashMap<OperationType, DecayMovingAverage>,
+    /// Counter for consecutive failures - resets on success
+    pub consecutive_failures: u32,
+    /// Time when validator was temporarily excluded due to failures.
+    /// Validators are excluded when consecutive_failures >= max_consecutive_failures
+    pub exclusion_time: Option<Instant>,
+}
+
+impl ValidatorClientStats {
+    pub fn new(init_reliability: f64) -> Self {
+        Self {
+            reliability: DecayMovingAverage::new(init_reliability, RELIABILITY_DECAY_FACTOR),
+            average_latencies: HashMap::new(),
+            consecutive_failures: 0,
+            exclusion_time: None,
+        }
+    }
+
+    pub fn update_average_latency(&mut self, operation: OperationType, new_latency: Duration) {
+        match self.average_latencies.entry(operation) {
+            Entry::Occupied(mut entry) => {
+                entry
+                    .get_mut()
+                    .update_moving_average(new_latency.as_secs_f64());
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(DecayMovingAverage::new(
+                    new_latency.as_secs_f64(),
+                    LATENCY_DECAY_FACTOR,
+                ));
+            }
+        }
+    }
+}
+
+/// Global statistics across all validators.
+///
+/// Used to track network-wide performance metrics that serve as baselines
+/// for scoring individual validators. Currently tracks maximum latencies
+/// for normalization purposes.
+#[derive(Debug, Clone, Default)]
+pub struct GlobalStats {
```

Comment:
Could error rate be useful here as well? I am thinking of an error scenario like overload where maybe a network wide metric might help dictate that we need to be more lenient about failures from individual validators